vllm/attention/ops/nki_blocksparse_flash_attn.py

```diff
@@ -0,0 +1,669 @@
+                self.tile_q_seq_ids,
+                self.tile_kv_seq_ids,
+        ]:
+            assert isinstance(arg,
+                              np.ndarray) and arg.dtype == np.int32, type(arg)
+            assert arg.shape[0] == self.num_tiles, (arg.shape, self.num_tiles)
+
+    @property
+    def num_tiles(self):
+        return len(self.tile_q_indices)
+
+    def build_tile_masks(self):
+        tile_kv_seq_ids = self.tile_kv_seq_ids
+        B_P_SIZE = 128
+        num_tiles, tile_size_kv = tile_kv_seq_ids.shape
+        block_size = self.block_size
+        num_tiled_blocks = max(B_P_SIZE, tile_size_kv // block_size)
+        assert (
+            tile_size_kv % B_P_SIZE == 0 and tile_size_kv % block_size == 0
+        ), f"{tile_size_kv=} is not multiple of {B_P_SIZE=} and {block_size=}"
+        tiled_block_size = tile_size_kv // num_tiled_blocks
+        if tiled_block_size > 1:
+            # reorder mask is needed somewhere as long as tiled_block_size > 1
+            tile_kv_seq_ids = tile_kv_seq_ids.reshape((
+                num_tiles,
+                num_tiled_blocks // B_P_SIZE,
+                B_P_SIZE,
+                tiled_block_size,
+            ))
+            tile_kv_seq_ids = tile_kv_seq_ids.transpose(0, 1, 3, 2).reshape(
+                (num_tiles, tile_size_kv))
+        return np.expand_dims(self.tile_q_seq_ids,
+                              2) == np.expand_dims(tile_kv_seq_ids, 1)
+
+    def build_tile_block_tables(self, block_tables):
+        tile_size_kv = self.tile_kv_seq_ids.shape[1]
+        num_blocks_per_tile = tile_size_kv // self.block_size
+        assert (tile_size_kv % self.block_size == 0
+                ), f"{tile_size_kv=} is not multiple of {self.block_size=}"
+        block_tables = block_tables.squeeze()
+        in_tile_offset = np.arange(num_blocks_per_tile)
+        indices = self.tile_block_table_offsets.reshape(
+            -1, 1) + in_tile_offset.reshape(1, -1)
+        return block_tables[indices]
+
+
+def _check_np_int_array(*arrays):
+    for a in arrays:
+        if not isinstance(a, np.ndarray) or a.dtype not in (np.int32,
+                                                            np.int64):
+            return False
+    return True
+
+
+class FlashAttentionPlanner:
+
+    def __init__(self, prompt_lens, context_lens, tile_size_q, tile_size_kv,
+                 block_size):
+        assert _check_np_int_array(prompt_lens, context_lens)
+        assert len(prompt_lens) == len(
+            context_lens
+        ), "prompt_lens and context_lens must have the same length"
+        self.num_seq = len(prompt_lens)
+        assert self.num_seq > 0, "prompt_lens and context_lens can't be empty"
+        self.prompt_lens = prompt_lens.astype(np.int32)
+        self.context_lens = context_lens.astype(np.int32)
+        self.tile_size_q = tile_size_q
+        self.tile_size_kv = tile_size_kv
+        self.block_size = block_size
+
+    def plan(self):
```

Comment:
Questions:
1. I suppose this is meant to be amortized for entire forward pass?  
2. The logic is quite involved, so I wonder what is the range of wallclock time for this function call for reasonable batch size (e.g. 128) - is it in us range?